File: Two_Pointers/015.3Sum/015.3Sum.java
class Solution {
  public List<List<Integer>> threeSum(int[] nums) {
  List < List < Integer >> ret = new ArrayList<>();
  Arrays.sort(nums);
  for (int i = 0; i < nums.length; i++) {

            int left = i + 1;
            int right = nums.length - 1;
            int sum = -nums[i];

    while (left < right) {
      if (nums[left] + nums[right] == sum) {
        ret.add(Arrays.asList(nums[i], nums[left], nums[right]));
        left++;
        right--;
        while (left < right && nums[left] == nums[left - 1]) {
          left++;
        }

        while (left < right && nums[right] == nums[right + 1]) {
          right--;
        }
      } else if (nums[left] + nums[right] < sum) {
        left++;
      } else {
        right--;
      }
    }
    while (i + 1 < nums.length && nums[i + 1] == nums[i]) {
      i++;
    }
  }
  return ret;
}
}

File: Two_Pointers/015.3Sum/015.3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ret;
        std::sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size() - 2; i++) {
            int j = i + 1;
            int k = nums.size() - 1;

            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum == 0) {
                    ret.push_back({ nums[i], nums[j], nums[k] });

                    while (j < k && nums[j] == nums[j + 1]) {
                        j++;
                    }

                    while (j < k && nums[k] == nums[k - 1]) {
                        k--;
                    }

                    j++;
                    k--;
                } else if (sum < 0) {
                    j++;
                } else {
                    k--;
                }
            }

            while (i + 1 < nums.size() && nums[i+1] == nums[i]) {
                i += 1;
            }
        }

        return ret;
    }
};

